File: shuffle/celeborn/client/src/main/scala/org/apache/celeborn/client/ShuffleTaskInfo.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.celeborn.client

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.function
import javax.annotation.concurrent.ThreadSafe

@ThreadSafe
class ShuffleTaskInfo {

  private var currentShuffleIndex: Int = 0
  // task shuffle id -> mapId_taskAttemptId -> attemptIdx
  private val taskShuffleAttemptIdToAttemptId =
    new ConcurrentHashMap[String, ConcurrentHashMap[String, Int]]
  // map attemptId index
  private val taskShuffleAttemptIdIndex =
    new ConcurrentHashMap[String, ConcurrentHashMap[Int, Int]]
  // task shuffle id -> celeborn shuffle id
  private val taskShuffleIdToShuffleId = new ConcurrentHashMap[String, Int]
  // celeborn shuffle id -> task shuffle id
  private val shuffleIdToTaskShuffleId = new ConcurrentHashMap[Int, String]

  val newMapFunc: function.Function[String, ConcurrentHashMap[Int, Int]] =
    new function.Function[String, ConcurrentHashMap[Int, Int]]() {
      override def apply(s: String): ConcurrentHashMap[Int, Int] = {
        new ConcurrentHashMap[Int, Int]()
      }
    }

  val newMapFunc2: function.Function[String, ConcurrentHashMap[String, Int]] =
    new function.Function[String, ConcurrentHashMap[String, Int]]() {
      override def apply(s: String): ConcurrentHashMap[String, Int] = {
        new ConcurrentHashMap[String, Int]()
      }
    }

  def getShuffleId(taskShuffleId: String): Int = {
    taskShuffleIdToShuffleId.synchronized {
      if (taskShuffleIdToShuffleId.containsKey(taskShuffleId)) {
        taskShuffleIdToShuffleId.get(taskShuffleId)
      } else {
        taskShuffleIdToShuffleId.put(taskShuffleId, currentShuffleIndex)
        shuffleIdToTaskShuffleId.put(currentShuffleIndex, taskShuffleId)
        val tempShuffleIndex = currentShuffleIndex
        currentShuffleIndex = currentShuffleIndex + 1
        tempShuffleIndex
      }
    }
  }

  def getAttemptId(taskShuffleId: String, mapId: Int, attemptId: String): Int = {
    val attemptIndex = taskShuffleAttemptIdIndex.computeIfAbsent(taskShuffleId, newMapFunc)
    val attemptIdMap =
      taskShuffleAttemptIdToAttemptId.computeIfAbsent(taskShuffleId, newMapFunc2)
    val mapAttemptId = mapId + "_" + attemptId
    attemptIndex.synchronized {
      if (!attemptIdMap.containsKey(mapAttemptId)) {
        if (attemptIndex.containsKey(mapId)) {
          val index = attemptIndex.get(mapId)
          attemptIdMap.put(mapAttemptId, index + 1)
          attemptIndex.put(mapId, index + 1)
        } else {
          attemptIdMap.put(mapAttemptId, 0)
          attemptIndex.put(mapId, 0)
        }
      }
    }

    attemptIdMap.get(mapAttemptId)
  }

  def removeExpiredShuffle(shuffleId: Int): Unit = {
    if (shuffleIdToTaskShuffleId.containsKey(shuffleId)) {
      val taskShuffleId = shuffleIdToTaskShuffleId.remove(shuffleId)
      taskShuffleIdToShuffleId.remove(taskShuffleId)
      taskShuffleAttemptIdIndex.remove(shuffleId)
      taskShuffleAttemptIdToAttemptId.remove(taskShuffleId)
    }
  }
}

File: shuffle/celeborn/client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/ShuffleTaskInfo.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.celeborn.plugin.flink;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.apache.celeborn.common.util.JavaUtils;

public class ShuffleTaskInfo {
  private static final Logger LOG = LoggerFactory.getLogger(ShuffleTaskInfo.class);

  private int currentShuffleIndex = 0;
  // map attemptId index
  private ConcurrentHashMap<Integer, ConcurrentHashMap<Integer, AtomicInteger>>
      shuffleIdMapAttemptIdIndex = JavaUtils.newConcurrentHashMap();
  // task shuffle id -> celeborn shuffle id
  private ConcurrentHashMap<String, Integer> taskShuffleIdToShuffleId =
      JavaUtils.newConcurrentHashMap();
  // celeborn shuffle id -> task shuffle id
  private ConcurrentHashMap<Integer, String> shuffleIdToTaskShuffleId =
      JavaUtils.newConcurrentHashMap();

  private ConcurrentHashMap<Integer, AtomicInteger> shuffleIdPartitionIdIndex =
      JavaUtils.newConcurrentHashMap();

  public int getShuffleId(String taskShuffleId) {
    synchronized (taskShuffleIdToShuffleId) {
      if (taskShuffleIdToShuffleId.containsKey(taskShuffleId)) {
        return taskShuffleIdToShuffleId.get(taskShuffleId);
      } else {
        taskShuffleIdToShuffleId.put(taskShuffleId, currentShuffleIndex);
        shuffleIdToTaskShuffleId.put(currentShuffleIndex, taskShuffleId);
        shuffleIdMapAttemptIdIndex.put(currentShuffleIndex, JavaUtils.newConcurrentHashMap());
        shuffleIdPartitionIdIndex.put(currentShuffleIndex, new AtomicInteger(0));
        int tempShuffleIndex = currentShuffleIndex;
        currentShuffleIndex = currentShuffleIndex + 1;
        return tempShuffleIndex;
      }
    }
  }

  public int genAttemptId(int shuffleId, int mapId) {
    AtomicInteger currentAttemptIndex =
        shuffleIdMapAttemptIdIndex
            .get(shuffleId)
            .computeIfAbsent(mapId, (id) -> new AtomicInteger(0));
    return currentAttemptIndex.getAndIncrement();
  }

  public int genPartitionId(int shuffleId) {
    return shuffleIdPartitionIdIndex.get(shuffleId).getAndIncrement();
  }

  public void removeExpiredShuffle(int shuffleId) {
    if (shuffleIdToTaskShuffleId.containsKey(shuffleId)) {
      shuffleIdPartitionIdIndex.remove(shuffleId);
      shuffleIdMapAttemptIdIndex.remove(shuffleId);
      String taskShuffleId = shuffleIdToTaskShuffleId.remove(shuffleId);
      taskShuffleIdToShuffleId.remove(taskShuffleId);
    }
  }

  public ShuffleResourceDescriptor genShuffleResourceDescriptor(
      String taskShuffleId, int mapId, String taskAttemptId) {
    int shuffleId = this.getShuffleId(taskShuffleId);
    int attemptId = this.genAttemptId(shuffleId, mapId);
    int partitionId = this.genPartitionId(shuffleId);
    LOG.info(
        "Assign for ({}, {}, {}) resource ({}, {}, {}, {})",
        taskShuffleId,
        mapId,
        taskAttemptId,
        shuffleId,
        mapId,
        attemptId,
        partitionId);
    return new ShuffleResourceDescriptor(shuffleId, mapId, attemptId, partitionId);
  }
}
